File: kirun-js/src/engine/function/system/array/Shuffle.ts
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { AbstractArrayFunction } from './AbstractArrayFunction';

export class Shuffle extends AbstractArrayFunction {
    public constructor() {
        super('Shuffle', [Shuffle.PARAMETER_ARRAY_SOURCE], Shuffle.EVENT_RESULT_EMPTY);
    }

    protected internalExecute(context: FunctionExecutionParameters): FunctionOutput {
        let source: any[] = context
            ?.getArguments()
            ?.get(Shuffle.PARAMETER_ARRAY_SOURCE.getParameterName());

        if (source.length <= 1) return new FunctionOutput([EventResult.outputOf(new Map([]))]);

        let x: number = 0;
        let size: number = source.length;

        for (let i: number = 0; i < size; i++) {
            let y: number = Math.floor(Math.random() * size) % size;
            let temp: any = source[x];
            source[x] = source[y];
            source[y] = temp;
            x = y;
        }

        return new FunctionOutput([EventResult.outputOf(new Map([]))]);
    }
}

File: kirun-py/src/kirun_py/function/system/array/shuffle.py
from __future__ import annotations

import random
from typing import Any, List, TYPE_CHECKING

from kirun_py.function.system.array.abstract_array_function import AbstractArrayFunction
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters


class Shuffle(AbstractArrayFunction):

    def __init__(self) -> None:
        super().__init__(
            'Shuffle',
            [AbstractArrayFunction.PARAMETER_ARRAY_SOURCE],
            AbstractArrayFunction.EVENT_RESULT_ARRAY,
        )

    async def internal_execute(
        self, context: FunctionExecutionParameters
    ) -> FunctionOutput:
        source: List[Any] = context.get_arguments().get(
            AbstractArrayFunction.PARAMETER_ARRAY_SOURCE.get_parameter_name()
        )

        if len(source) <= 1:
            return FunctionOutput([
                EventResult.output_of(
                    {AbstractArrayFunction.EVENT_RESULT_NAME: source}
                )
            ])

        source = list(source)
        size: int = len(source)
        x: int = 0

        for i in range(size):
            y: int = random.randint(0, size - 1)
            source[x], source[y] = source[y], source[x]
            x = y

        return FunctionOutput([
            EventResult.output_of(
                {AbstractArrayFunction.EVENT_RESULT_NAME: source}
            )
        ])
